File: 백준/Silver/11403. 경로 찾기/경로 찾기.cc
#include <iostream>
#include <vector>
#include <stack>
#include <cmath>
#include <queue>
#include <map>
#include<unordered_map>
#include<unordered_set>
#include <algorithm>
#include <cstring>
#include <string>
#include <sstream>
#include <deque>
using namespace std;

int N;
int Map[101][101];
int Visit[101];
vector<int> Vector[101];



void Dfs(int x) {
	for (int i = 0; i < Vector[x].size(); i++)
	{
		int next = Vector[x][i];
		if (Visit[next] == 1) continue;
		Visit[next] = 1;
		Dfs(next);
	}
}


int main() {
	
	cin >> N;
	for (int i = 1; i <= N; i++)
		for (int j = 1; j <= N; j++)
		{
			cin >> Map[i][j];
			if (Map[i][j] == 1)
				Vector[i].push_back(j);
		}
		

	for (int i = 1; i <= N; i++)
	{
		Dfs(i);
		for (int j = 1; j <= N; j++)
			if (Visit[j] == 1)
				Map[i][j] = 1;

		memset(Visit, 0, sizeof(Visit));
	}


	for (int i = 1; i <= N; i++)
	{
		for (int j = 1; j <= N; j++)
		{
			cout << Map[i][j] << ' ';
		}
		cout << endl;
	}
}

File: 백준/Silver/11403. 경로 찾기/경로 찾기.java
import java.util.ArrayList;
import java.util.Scanner;

public class Main {

    static int N;
    static int[][] map;
    static int[] visit;
    static ArrayList<Integer>[] Arr;

    public static void dfs(int x) {
        for (int next : Arr[x]) {
            if (visit[next] == 1) continue;
            visit[next] = 1;
            dfs(next);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        N = scanner.nextInt();
        map = new int[N + 1][N + 1];
        visit = new int[N + 1];
        Arr = new ArrayList[N + 1];

        for (int i = 1; i <= N; i++) {
            
            Arr[i] = new ArrayList<>();
            
            for (int j = 1; j <= N; j++) {
                map[i][j] = scanner.nextInt();
                if (map[i][j] == 1) {
                    Arr[i].add(j);
                }
            }
        }

        for (int i = 1; i <= N; i++) {
            dfs(i);
            for (int j = 1; j <= N; j++) {
                if (visit[j] == 1) {
                    map[i][j] = 1;
                }
                visit[j] = 0;
            }
        }

        for (int i = 1; i <= N; i++) {
            for (int j = 1; j <= N; j++) {
                System.out.print(map[i][j] + " ");
            }
            System.out.println();
        }
    }
}



